oeps/oep-0039-yaml-config.rst

```diff
@@ -0,0 +1,146 @@
++-----------------+--------------------------------------------------------+
+| Arbiter         | TBA                                                    |
++-----------------+--------------------------------------------------------+
+| Status          | Draft                                                  |
++-----------------+--------------------------------------------------------+
+| Type            | Best Practice                                          |
++-----------------+--------------------------------------------------------+
+| Created         | 2019-07-15                                             |
++-----------------+--------------------------------------------------------+
+| `Review Period` | <start - target end dates for review>                  |
++-----------------+--------------------------------------------------------+
+| `Resolution`    |                                                        |
++-----------------+--------------------------------------------------------+
+| `References`    |                                                        |
++-----------------+--------------------------------------------------------+
+
+Abstract
+========
+
+Yaml configuration instead of python django configuration settings files for configuring django applications is preferable because it simplifies configuration of django applications
+
+Motivation
+==========
+
+Currently there is no standard for this and many applications have varying takes on configuration. 
+Since Open edX consists of many applications managing and configuring it can be difficult, It would be much easier if all of the applications were configured consistently. 
+This OEP intends to provide a set of standard best practices for configuring Open edX in a python django application.
+
+Specification
+=============
+
+Your configuration should consist of two files
+
+settings.py
+defaults.py 
+
+defaults.py should list every configuration key in the application, and provide a description of what it does.
+
+settings.py should do two things:
+1) It should first load defaults.py into the python settings module.
+2) Then should load a single yaml file into the default keyspace overwriting values in the settings module.
+
+The yaml file path loaded should be configurable via environment variable that is specific to the application
+eg: ECOMMERCE_CFG. 
+
+There should be no additional django configuration settings files like devstack.py, testing.py, docker.py, docker_devstack.py, or anything of the sort.
+
+Plugins should not modify the loaded settings python module. Since we load the keys from the yaml file, even the ones we do not know at the application layer.
+A plugin should not need to modify the configuration of the application.
+
+All of the configuration should be done in the yaml file
+
+configuration should never be appended to or removed from in python code after settings.py is loaded, only read.
+
+eg:
```

Comment:
Currently in the edxapp config we append to lists of keys in the config after it is loaded to avoid having to overwrite the key and specify the defaults again, this is done in a few places

here is an example of one:
https://github.com/edx/edx-platform/blob/master/lms/envs/production.py#L143

Basically it is an extension of config should be read and not modified.

Reply:
The intent of it was just to clarify that the yml config key
MY_KEY

should be accessible as settings.MY_KEY in the python code.

I will clarify the example

Reply:
Here is another good example(that is currently biting me):
`/home/coryleeio/Workspace/edx-platform/lms/envs/bok_choy.py:
  254  # Make sure we test with the extended history table
  255  FEATURES['ENABLE_CSMH_EXTENDED'] = True
  256: INSTALLED_APPS.append('coursewarehistoryextended')
  257  
  258  BADGING_BACKEND = 'lms.djangoapps.badges.backends.tests.dummy_backend.DummyBackend'

/home/coryleeio/Workspace/edx-platform/lms/envs/production.py:
  929  # The extended StudentModule history table
  930  if FEATURES.get('ENABLE_CSMH_EXTENDED'):
  931:     INSTALLED_APPS.append('coursewarehistoryextended')
  932  
  933  API_ACCESS_MANAGER_EMAIL = ENV_TOKENS.get('API_ACCESS_MANAGER_EMAIL')

/home/coryleeio/Workspace/edx-platform/lms/envs/test.py:
  207  # Make sure we test with the extended history table
  208  FEATURES['ENABLE_CSMH_EXTENDED'] = True
  209: INSTALLED_APPS.append('coursewarehistoryextended')`